codex/sales/states/filled.nim

```diff
@@ -0,0 +1,46 @@
+import pkg/questionable
+import ./errored
+import ./finished
+import ./cancelled
+import ./failed
+import ../statemachine
+
+type
+  SaleFilled* = ref object of SaleState
+  SaleFilledError* = object of CatchableError
+
+method onCancelled*(state: SaleFilled, request: StorageRequest) {.async.} =
+  await state.switchAsync(SaleCancelled())
+
+method onFailed*(state: SaleFilled, request: StorageRequest) {.async.} =
+  await state.switchAsync(SaleFailed())
+
+method `$`*(state: SaleFilled): string = "SaleFilled"
+
+method enterAsync(state: SaleFilled) {.async.} =
+  without agent =? (state.context as SalesAgent):
+    raiseAssert "invalid state"
+
+  try:
+    let market = agent.sales.market
+
+    without slotIndex =? agent.slotIndex:
+      raiseAssert "no slot selected"
+
+    if availability =? agent.availability:
+      agent.sales.remove(availability)
+
+    let host = await market.getHost(agent.requestId, slotIndex)
+    let me = await market.getSigner()
+    if host == me.some:
+      await state.switchAsync(SaleFinished())
+    else:
+      let error = newException(SaleFilledError, "Sale host mismatch")
```

Comment:
This is not really an error as such? We simply lost the slot to another host that was faster.

Reply:
I agree that an exception may be too harsh for this situation, but it is still an graceful error of sorts as it is an unexpected outcome. Maybe we could create a new property on the `SaleErrored` state that indicates a graceful error?

Reply:
You're right, let's keep it like it is.